va/va.go

```diff
@@ -29,96 +39,177 @@ func userAgent() string {
+	Registration *core.Registration
+	Results      chan core.ValidationRecord
+}
+
 type VAImpl struct {
 	log      *log.Logger
 	clk      clock.Clock
 	httpPort int
+	tasks    chan *vaTask
+	ca       *ca.CAImpl
 }
 
-func New(log *log.Logger, clk clock.Clock, httpPort int) *VAImpl {
-	return &VAImpl{
+func New(log *log.Logger, clk clock.Clock, httpPort int, ca *ca.CAImpl) *VAImpl {
+	va := &VAImpl{
 		log:      log,
 		clk:      clk,
 		httpPort: httpPort,
+		tasks:    make(chan *vaTask, taskQueueSize),
+		ca:       ca,
 	}
+
+	go va.processTasks()
+	return va
 }
 
-func (va VAImpl) Validate(identifier string, chal *core.Challenge, reg *core.Registration) error {
-	// TODO(@cpu): Implement validation for DNS-01, TLS-SNI-02, etc
-	var prob *acme.ProblemDetails
-	switch chal.Type {
-	case acme.ChallengeHTTP01:
-		prob = va.validateHTTP01(identifier, chal, reg)
-	default:
-		return fmt.Errorf("Invalid challenge type: %q", chal.Type)
+func (va VAImpl) ValidateChallenge(ident string, chal *core.Challenge, reg *core.Registration) {
+	task := &vaTask{
+		Identifier:   ident,
+		Challenge:    chal,
+		Registration: reg,
+		Results:      make(chan core.ValidationRecord, concurrentValidations),
 	}
+	// Submit the task for validation
+	va.tasks <- task
+}
 
-	authz := chal.Authz
-	now := va.clk.Now().UTC()
-	// Update the validated date for the challenge regardless of good/bad outcome
-	chal.ValidatedDate = now
-	chal.Validated = chal.ValidatedDate.Format(time.RFC3339)
-	if prob != nil {
-		// Update the challenge Error
-		chal.Error = prob
-		// Set the challenge and authorization to invalid
-		chal.Status = acme.StatusInvalid
-		authz.Status = acme.StatusInvalid
-		va.log.Printf("authz %s set INVALID by completed challenge %s", authz.ID, chal.ID)
-	} else {
+func (va VAImpl) processTasks() {
+TaskLoop:
+	for {
+		// Pull a task off of the tasks channel
+		task, ok := <-va.tasks
+		if !ok {
+			break
+		}
+		va.log.Printf("Pulled a task from the Tasks queue: %#v", task)
+		va.log.Printf("Starting %d validations.", concurrentValidations)
+
+		var wg sync.WaitGroup
+		wg.Add(concurrentValidations)
+		// Start a number of go routines to perform concurrent validations
+		for i := 0; i < concurrentValidations; i++ {
+			go va.performValidation(task, &wg)
+		}
+		va.log.Printf("Waiting on validations")
+		// Wait for all of the go routines to finish
+		wg.Wait()
+		va.log.Printf("All %d validations finished.\n", concurrentValidations)
+
+		chal := task.Challenge
+		authz := chal.Authz
+
+		// Update the validated date for the challenge regardless of good/bad outcome
+		now := va.clk.Now().UTC()
+		chal.ValidatedDate = now
+		chal.Validated = chal.ValidatedDate.Format(time.RFC3339)
+
+		// Read a validation result from the task results channel for each of the
+		// concurrent validations into a slice for processing.
+		for i := 0; i < concurrentValidations; i++ {
+			va.log.Printf("Reading a result from task.Results")
+			result := <-task.Results
```

Comment:
Since we wait on the WaitGroup above, `task.Results` is guaranteed to be exactly filled, and there is guaranteed to be no concurrent access. So there's no reason to make it a channel. It can just be a slice, and then you can `range` over it. Alternately you could get rid of the WaitGroup and use just the channel, which would allow you to surface errors without waiting for all validations to come in. Basically, we need just one of these concurrency primitives here.

Also, I'd recommend factoring this out into a `firstError` function or similar, that returns either an error or nil. That way you can call that function and have an if/else that very clearly handles both the error and non-error cases in one place.